pkg/distributor/influx.go

```diff
@@ -0,0 +1,177 @@
+		ctx := r.Context()
+		logger := utillog.WithContext(ctx, logger)
+		if sourceIPs != nil {
+			source := sourceIPs.Get(r)
+			if source != "" {
+				logger = utillog.WithSourceIPs(source, logger)
+			}
+		}
+
+		supplier := func() (*mimirpb.WriteRequest, func(), error) {
+			rb := util.NewRequestBuffers(requestBufferPool)
+			var req mimirpb.PreallocWriteRequest
+
+			if err := parser(ctx, r, maxRecvMsgSize, rb, &req, logger); err != nil {
+				// TODO(alexg): Do we even need httpgrpc here?
+				// Check for httpgrpc error, default to client error if parsing failed
+				if _, ok := httpgrpc.HTTPResponseFromError(err); !ok {
+					err = httpgrpc.Error(http.StatusBadRequest, err.Error())
+				}
+
+				rb.CleanUp()
+				return nil, nil, err
+			}
+
+			cleanup := func() {
+				mimirpb.ReuseSlice(req.Timeseries)
+				rb.CleanUp()
+			}
+			return &req.WriteRequest, cleanup, nil
+		}
+		req := newRequest(supplier)
+		// https://docs.influxdata.com/influxdb/cloud/api/v2/#tag/Response-codes
+		if err := push(ctx, req); err != nil {
+			if errors.Is(err, context.Canceled) {
+				level.Warn(logger).Log("msg", "push request canceled", "err", err)
+				w.WriteHeader(statusClientClosedRequest)
+				return
+			}
+			if errors.Is(err, io2.ErrReadLimitExceeded) {
+				// TODO(alexg): One thing we have seen in the past is that telegraf clients send a batch of data
+				// if it is too big they should respond to the 413 below, but if a client doesn't understand this
+				// it just sends the next batch that is even bigger. In the past this has had to be dealt with by
+				// adding rate limits to drop the payloads.
+				level.Warn(logger).Log("msg", "request too large", "err", err)
+				// TODO(alexg): max size and bytes received in error?
+				w.WriteHeader(http.StatusRequestEntityTooLarge)
+				return
+			}
+			// From: https://github.com/grafana/influx2cortex/blob/main/pkg/influx/errors.go
+
+			var (
+				httpCode int
+				errorMsg string
+			)
+			if st, ok := grpcutil.ErrorToStatus(err); ok {
+				// TODO(alexg): Hmm, still needed?
+				// This code is needed for a correct handling of errors returned by the supplier function.
+				// These errors are created by using the httpgrpc package.
+				httpCode = int(st.Code())
+				errorMsg = st.Message()
+			} else {
+				var distributorErr Error
+				errorMsg = err.Error()
+				if errors.Is(err, context.DeadlineExceeded) || !errors.As(err, &distributorErr) {
+					httpCode = http.StatusServiceUnavailable
+				} else {
+					httpCode = errorCauseToHTTPStatusCode(distributorErr.Cause(), false)
+				}
+			}
+			if httpCode != 202 {
+				// This error message is consistent with error message in Prometheus remote-write handler, and ingester's ingest-storage pushToStorage method.
+				msgs := []interface{}{"msg", "detected an error while ingesting Influx metrics request (the request may have been partially ingested)", "httpCode", httpCode, "err", err}
+				if httpCode/100 == 4 {
+					// TODO(alexg): what is this?
```

Comment:
Logs with `insight=true` are visible to our Grafana Cloud customers.

Reply:
Aha, TIL! Will add it back in as that could be useful.